File: sdk/dotnet/src/Cordum.Cap/Metrics.cs
namespace Cordum.Cap;

/// <summary>
/// Pluggable observability callbacks for CAP protocol events.
/// All methods have empty default implementations.
/// </summary>
public interface IMetricsHook
{
    void OnJobReceived(string jobId, string topic) { }
    void OnJobCompleted(string jobId, long durationMs, string status) { }
    void OnJobFailed(string jobId, string errorMsg) { }
    void OnHeartbeatSent(string workerId) { }
    void OnProgressEmitted(string jobId, int percent) { }
    void OnError(string context, string errorMsg) { }
}

/// <summary>
/// No-op metrics implementation. All callbacks are empty.
/// </summary>
public sealed class NoopMetrics : IMetricsHook { }

File: sdk/java/src/main/java/io/cordum/cap/Metrics.java
package io.cordum.cap;

public interface Metrics {

    void onJobReceived(String jobId, String topic);

    void onJobCompleted(String jobId, long durationMs, String status);

    void onJobFailed(String jobId, String errorMsg);

    void onHeartbeatSent(String workerId);

    void onProgressEmitted(String jobId, int percent);

    void onError(String context, String errorMsg);

    Metrics NOOP = new Metrics() {
        @Override public void onJobReceived(String jobId, String topic) {}
        @Override public void onJobCompleted(String jobId, long durationMs, String status) {}
        @Override public void onJobFailed(String jobId, String errorMsg) {}
        @Override public void onHeartbeatSent(String workerId) {}
        @Override public void onProgressEmitted(String jobId, int percent) {}
        @Override public void onError(String context, String errorMsg) {}
    };
}
